autoretrieve/linksystem.go

```diff
@@ -0,0 +1,181 @@
+package autoretrieve
+
+import (
+	"bytes"
+	"errors"
+	"io"
+
+	provider "github.com/filecoin-project/index-provider"
+	"github.com/filecoin-project/storetheindex/api/v0/ingest/schema"
+	"github.com/ipfs/go-datastore"
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/codec/dagjson"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/ipld/go-ipld-prime/node/basicnode"
+)
+
+var errNoEntries = errors.New("no entries; see schema.NoEntries")
+
+// Creates the main engine linksystem.
+func (e *AutoretrieveEngine) mkLinkSystem() ipld.LinkSystem {
+	lsys := cidlink.DefaultLinkSystem()
+	lsys.StorageReadOpener = func(lctx ipld.LinkContext, lnk ipld.Link) (io.Reader, error) {
+		// If link corresponds to schema.NoEntries return error immediately.
+		if lnk == schema.NoEntries {
+			return nil, errNoEntries
+		}
+
+		ctx := lctx.Ctx
+		c := lnk.(cidlink.Link).Cid
+		log.Debugf("Triggered ReadOpener from engine's linksystem with cid (%s)", c)
+
+		// Get the node from main datastore. If it is in the
+		// main datastore it means it is an advertisement.
+		val, err := e.ds.Get(ctx, datastore.NewKey(c.String()))
```

Comment:
are you not supposed to be using the `dsLatestAdvKey` or calling `e.getLatestAdCid` here?